flecsi/run/charm/context.cc

```diff
@@ -0,0 +1,123 @@
+/*
+    @@@@@@@@  @@           @@@@@@   @@@@@@@@ @@
+   /@@/////  /@@          @@////@@ @@////// /@@
+   /@@       /@@  @@@@@  @@    // /@@       /@@
+   /@@@@@@@  /@@ @@///@@/@@       /@@@@@@@@@/@@
+   /@@////   /@@/@@@@@@@/@@       ////////@@/@@
+   /@@       /@@/@@//// //@@    @@       /@@/@@
+   /@@       @@@//@@@@@@ //@@@@@@  @@@@@@@@ /@@
+   //       ///  //////   //////  ////////  //
+
+   Copyright (c) 2016, Triad National Security, LLC
+   All rights reserved.
+                                                                              */
+
+#include <flecsi-config.h>
+
+#if !defined(__FLECSI_PRIVATE__)
+#define __FLECSI_PRIVATE__
+#endif
+
+#include "flecsi/exec/launch.hh"
+#include "flecsi/exec/charm/task_wrapper.hh"
+#include "flecsi/run/charm/context.hh"
+#include "flecsi/run/types.hh"
+#include <flecsi/data.hh>
+
+#include <mpi-interoperate.h>
+
+#include "context.def.h"
+
+namespace flecsi::run {
+
+using namespace boost::program_options;
+using exec::charm::task_id;
+
+namespace charm {
+
+// TODO: Depth set to -1 because it's being incremented somewhere before
+// the top level action executes.
+ContextGroup::ContextGroup() : depth(-1) {
+  if (CkMyPe() != 0) {
+    run::context::instance().context_proxy_ = thisProxy;
+  }
+}
+
+void ContextGroup::top_level_task() {
+  context_t & context_ = context_t::instance();
+  detail::data_guard(),
+    context_.exit_status() = (*context_.top_level_action_)();
+  if (CkMyPe() == 0) {
+    CkStartQD(CkCallback(CkCallback::ckExit));
+  }
+}
+}
+
+//----------------------------------------------------------------------------//
+// Implementation of context_t::initialize.
+//----------------------------------------------------------------------------//
+
+int
+context_t::initialize(int argc, char ** argv, bool dependent) {
+
+  if(dependent) {
+    CharmBeginInit(argc, argv);
+    if (CkMyPe() == 0) {
+      context_proxy_ = charm::CProxy_ContextGroup::ckNew();
+    }
+    CharmFinishInit();
+  } // if
+
+  context::process_ = CkMyPe();
+  context::processes_ = CkNumPes();
+
+  auto status = context::initialize_generic(argc, argv, dependent);
+
+  if(status != success && dependent) {
+    CharmLibExit();
+  } // if
+
+  return status;
+} // initialize
+
+//----------------------------------------------------------------------------//
+// Implementation of context_t::finalize.
+//----------------------------------------------------------------------------//
+
+void
+context_t::finalize() {
+  context::finalize_generic();
+
+  if(context::initialize_dependent_) {
+    CharmLibExit();
+    MPI_Finalize();
```

Comment:
This actually needs to change. We made some small changes to the Charm interop interfaces while working with Flecsi. CharmInit calls MPI_Init, and CharmLibExit() calls MPI_Finalize, so I'll need to remove the extra call to MPI_Finalize() here.

Reply:
I thought it might, given the `argc`/`argv` arguments.  Without claiming to know anything about Charm++, I wonder why you want that: obviously any two libraries that do so can't be used together.

Reply:
It depends on the mode you are using Charm++ in and the type of Charm++ build you are doing. The mode we have chosen for Flecsi is what we refer to as user-driven mode. The user calls CharmInit and CharmLibExit in basically the same way they would call MPI_Init/MPI_Finalize. Those calls initialize and finalize the underlying communication layers. If you build the MPI version of Charm++, then that would include initializing and finalizing MPI. However, if Flecsi doesn't have an inherent need for MPI, then you could build Charm++ on top of any communication layer you want.

If instead you are planning on explicit interoperation with Charm++ and MPI, you would then use the interop mode for Charm++. In that mode, you are responsible for initializing and finalizing MPI yourself, and when you initialize Charm++ you give it the MPI communicator you want it to live within. In this mode you can initialize as many charm instances as you want, and CharmLibExit() has no effect on your MPI environment.

Since Flecsi doesn't explicitly rely on MPI to our knowledge, we figured the user driven mode made the most sense. This extra MPI_Finalize should have been removed for sure, but I think I just missed it when preparing the pull request.

Reply:
Good to know.  While we don't assume that FleCSI clients will be absolutely portable between backends, it is worth noting that we expect clients that target Legion to occasionally use MPI directly (in so-called "MPI tasks": `flecsi::execute<foo,flecsi::mpi>(…)`).  Whether that means that they will care about initializing Charm++ themselves or using something other than `MPI_COMM_WORLD` for it, I can't say.

Reply:
Ah OK. That's certainly something the Charm backend could also support pretty easily. The initialize and finalize methods would then just call MPI_Init/MPI_Finalize like they do in Legion, as well as calling our interop CharmInit and CharmExit. Part of the reason we didn't do that initially is because that would restrict us to using a Charm that is built on top of MPI. Whereas in the current version, you can use Charm++ built on whatever communication layer makes the most sense for where you are running it.